common/js/src/executable-module/offscreen-doc-emscripten-module.js

```diff
@@ -0,0 +1,256 @@
+
+goog.require('GoogleSmartCard.ExecutableModule');
+goog.require('GoogleSmartCard.Logging');
+goog.require('GoogleSmartCard.PortMessageChannel');
+goog.require('GoogleSmartCard.PromiseHelpers');
+goog.require('goog.Promise');
+goog.require('goog.log');
+goog.require('goog.log.Logger');
+goog.require('goog.messaging.AbstractChannel');
+goog.require('goog.object');
+goog.require('goog.promise.Resolver');
+
+goog.scope(function() {
+
+const GSC = GoogleSmartCard;
+
+const OFFSCREEN_DOC_URL = '/emscripten-offscreen-doc.html';
+// See
+// <https://developer.chrome.com/docs/extensions/reference/api/offscreen#type-Reason>:
+const OFFSCREEN_DOC_REASONS = ['WORKERS'];
+const OFFSCREEN_DOC_JUSTIFICATION = 'running a WebAssembly module';
+const URL_PARAM = 'google_smart_card_emscripten_module_name';
+// The name of the port that the Offscreen Document should use for exchanging
+// the module's messages with us.
+const MESSAGING_PORT_NAME = 'offscreen_doc_emscripten_messages';
+// The name of the port that the Offscreen Document should use for signaling the
+// status.
+const STATUS_PORT_NAME = 'offscreen_doc_emscripten_status';
+// Status messages to be sent over the `STATUS_PORT_NAME` port.
+const STATUS_LOADED = 'loaded';
+const STATUS_DISPOSED = 'disposed';
+
+GSC.OffscreenDocEmscriptenModule = class extends GSC.ExecutableModule {
+  /**
+   * @param {string} moduleName
+   */
+  constructor(moduleName) {
+    super();
+    /** @type {string} @const @private */
+    this.moduleName_ = moduleName;
+    /** @type {!goog.log.Logger} @const @private */
+    this.logger_ = GSC.Logging.getScopedLogger('OffscreenDocEmscriptenModule');
+    /** @type {boolean} @private */
+    this.offscreenDocActive_ = false;
+    /** @type {!OffscreenDocMessageChannel} @const @private */
+    this.messageChannel_ = new OffscreenDocMessageChannel();
+    /** @type {!GSC.PortMessageChannel|null} @private */
+    this.statusChannel_ = null;
+    /** @type {!goog.promise.Resolver<void>} @const @private */
+    this.loadPromiseResolver_ = goog.Promise.withResolver();
+    GSC.PromiseHelpers.suppressUnhandledRejectionError(
+        this.loadPromiseResolver_.promise);
+  }
+
+  /** @override */
+  getLogger() {
+    return this.logger_;
+  }
+
+  /** @override */
+  startLoading() {
+    // Listen for incoming message channels from the Offscreen Document once
+    // it's loaded.
+    chrome.runtime.onConnect.addListener((port) => this.onConnect_(port));
+
+    // Load the Offscreen Document. The loading arguments are passed via URL
+    // query parameters.
+    const url = new URL(OFFSCREEN_DOC_URL, globalThis.location.href);
+    url.searchParams.append(URL_PARAM, this.moduleName_);
+    chrome.offscreen.createDocument(
+        {
+          'url': url.toString(),
+          'reasons': OFFSCREEN_DOC_REASONS,
+          'justification': OFFSCREEN_DOC_JUSTIFICATION,
+        },
+        () => this.onOffscreenDocCreated_());
```

Comment:
Most of SCC is relying on callbacks, but since this API is MV3 exclusive anyway, we could consider using promises instead here - IMO they make the code more readable.

Reply:
Done - the linear logic is definitely more readable; thanks for the suggestion!

The only downside is that we need to handle the scenario "what if we got disposed in the middle of this long async function", but it's not too bad I think.